### PR TITLE
Add hide_queryset_annotations helper

### PR DESCRIPTION
**Pull request by delfick**
It's useful to be able to tell mypy to ignore the queryset annotations when we've only added an annotation for a filter applied in that place.

For example when we want to return a non annotated queryset from a function, but require an annotation within that function to determine what's in the queryset

Currently when we do this we need to specify those annotations in the return type of such a function even when they aren't used outside that function

This is a response to https://github.com/typeddjango/django-stubs/issues/3382

---

AI Disclosure: **NONE**: I actively avoid LLMs: No LLM or any other machine learning technique or product was used during the ideation, creation or description of this change[.](https://ky.fyi/posts/ai-burnout)

### CODE REVIEW

File: docs/api/usage.rst [hunk 1]
```diff
@@ -193,3 +193,34 @@ This also works on the concrete models themselves:
...
+        return qs.annotate(max_value=Max("value")).filter(max_value__gt=10)
+
+In this case we can remove the queryset annotations:
+
+.. code-block:: python
+
+    from extended_django_mypy_plugin import hide_queryset_annotations
+
+    def some_filter(qs: MyQuerySet[MyModel]) -> MyQuerySet[MyModel]:
+        return hide_queryset_annotations
```
Comment (mitch104) [inline docs/api/usage.rst]: 🚗 Is this missing an open bracket?
```suggestion
        return hide_queryset_annotations(
```

Comment (delfick) [reply docs/api/usage.rst]: oops, thanks, fixed :)

File: extended_mypy_django_plugin/_plugin/type_checker.py [hunk 2]
```diff
@@ -20,6 +28,75 @@ class TypeChecking:
...
+            second_arg = get_proper_type(first_arg.args[1])
+            if isinstance(second_arg, TypeAliasType) and second_arg.alias:
+                second_arg = get_proper_type(second_arg.alias.target)
+
+            if isinstance(second_arg, TypedDictType):
+                queryset_args.append(first_arg.args[1])
+
+        return ctx.default_signature.copy_modified(
+            arg_types=[first_arg], ret_type=Instance(first_arg.type, queryset_args)
+        )
```
Comment (ngnpope) [inline extended_mypy_django_plugin/_plugin/type_checker.py]: I struggled a bit more when I got to this point. Perhaps some comments would help? As far as I could tell, this is hoisting the annotations from the `Model` in the first type arg to the second type arg of the `QuerySet`? But then I don't see a test to indicate that's what's happening...

**Review by ngnpope — COMMENTED** [review]
In general I think this looks fine and handles most edge cases - I guess you'll have tested this internally anyway.

A few thoughts: This seems simple enough to upstream - doesn't require the virtual stuff. Do you have any thoughts on that? I'm concerned that we might grow too many extras on top that other type checkers would then not implement when they get around to adding Django support.

In addition, is this approach the right way - a cast-like function? Or would we be better off with a `WithoutAnnotations` (although that would apply to the `Model`, e.g. `-> MyQuerySet[WithoutAnnotations[MyModel]]`) or something like `-> StripAnnotations[MyQuerySet[MyModel]]`. Totally not thought about this in depth - might be totally naive and non-viable. I was merely thinking about this being something that effectively allows a broader type.

File: extended_mypy_django_plugin/_plugin/type_checker.py (hunk 2 above)
Comment (delfick) [reply extended_mypy_django_plugin/_plugin/type_checker.py]: fair, I'll add some comments and get back to you explaining what that's about (it's the row type)

**Comment by delfick** [comment]
@ngnpope oh, 100%. I'm putting it in here first so I can use it to resolve the 100s of errors the django-stubs 6.0.5 upgrade is giving us. As I use it more widely and iron out anything about it I haven't considered I'll go back to that github issue and point this out and see if django-stubs wants to include their own version, and if that happens I'll remove it from here

As for being a cast. Doing it this way is necessary because I'm changing the inferred type. So if it was only an annotation, there's no way to convince mypy that it actually matches.

File: extended_mypy_django_plugin/_plugin/type_checker.py (hunk 2 above)
Comment (delfick) [reply extended_mypy_django_plugin/_plugin/type_checker.py]: @ngnpope now with comments! I also saw a way to simplify it slightly.

Ultimately we're protecting against this

```
qs = MyQuerySet.objects.all().annotate(...).values(...)
reveal_type(qs) # django.db.models.QuerySet[MyModel@Annotated(...), TypedDict(...)]

reveal_type(hide_queryset_annotations(qs)) # django.db.models.QuerySet[MyModel]
```

because having a values queryset is a different interface

**Review by jamesbeith — APPROVED** [review]
